core/jazz_events-handler/index.js

```diff
@@ -14,125 +14,135 @@
-								"sequence_id": sequenceNumber,
-								"event": payload,
-								"failure_code": null,
-								"failure_message": null
-							});
-							innerCallback(null, {
-								"message": "successfully processed interested event"
-							});
-						}
-
-					});
-				} else {
-					processedEvents.push({
-						"sequence_id": sequenceNumber,
-						"event": payload,
-						"failure_code": null,
-						"failure_message": null
-					});
-					logger.info('Skipping storage of un-interested event');
-					innerCallback(null, {
-						"message": "successfully processed un-interested event"
-					});
-				}
-
-			}, function (err) {
-				callback();
-			});
-
-		}
-
-	}, function (err, results) {
-		if (err) {
-			return logger.error('Error inside events handler ' + JSON.stringify(err));
-		}
-
-		cb(null, {
-			"processed_events": processedEvents.length,
-			"failed_events": failedEvents.length
-		});
-	});
-
+var processedEvents = [];
+var failedEvents = [];
+var failureCodes = fcodes();
+
+function handler(event, context, cb) {
+  //Initializations
+  let config = configModule.getConfig(event, context);
+
+  exportable.getEvents(config)
+    .then(res => exportable.processRecords(config, event, res))
+    .then(() => { 
+      return cb(null, { "processed_events": processedEvents.length, "failed_events": failedEvents.length }); 
+    })
+    .catch(err => {
+      return logger.error('Error inside events handler ' + JSON.stringify(err));
+    });
 };
+
+function getEvents(config) {
+  return new Promise((resolve, reject) => {
+    const docClient = new AWS.DynamoDB.DocumentClient();
+    let interestedEvents = [];
+    let params = {
+      TableName: config.EVENT_NAME_TABLE,
+      ProjectionExpression: "EVENT_NAME",
+      Limit: 1000,
+      ReturnConsumedCapacity: "TOTAL"
+    };
+
+    let scanExecute = function () {
+      docClient.scan(params, function (err, result) {
+        if (err) {
+          return reject({ "db_error": "Unable to scan Event Names table to fetch interested events" });
+        } else {
+          //interestedEvents = interestedEvents.concat(result.Items);
+          for (i = 0; i < result.Items.length; i++) {
+            interestedEvents.push(result.Items[i].EVENT_NAME);
+          }
+
+          if (result.LastEvaluatedKey) {
+            params.ExclusiveStartKey = result.LastEvaluatedKey;
+            scanExecute();
+          } else {
+            return resolve(interestedEvents);
+          }
+        }
+      });
+    };
+    scanExecute();
+  })
+}
+
+function processRecords(config, event, res) {
+  return new Promise((resolve, reject) => {    
+    let processEachEventPromises = [];
+    for (let i = 0; i < event.Records.length; i++) {
+      processEachEventPromises.push(exportable.processEachEvent(event.Records[i], config, res));
+    }
+    Promise.all(processEachEventPromises)
+      .then((result) => {
+        return resolve(result);
+      })
+      .catch((error) => {
+        return resolve();
```

Comment:
is it suppose to be resolved?

Reply:
in the original code, on error in async.each the callback is called with no params. So I replaced it with this code

Reply:
not sure how this is better. ideally it should be reject.

Reply:
The way the code is written, sending reject here would make the service not send any response back to the caller. AFAIK the intention of the author was to send the number of processed and failed events back to the caller.

Reply:
Not really, reject will send the response, its just that you need to have a catch block to handle it. The way code is written is not the right way of using promises. I know you have not written it so this can be a enhancement later.